### PR TITLE
[CT-3267] [Feature] Debug log when `type_code` fails to convert to a `data_type`

### LINKED ISSUE #38: [CT-3267] [Feature] Debug log when `type_code` fails to convert to a `data_type`
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

_[Originally](https://github.com/dbt-labs/dbt-core/issues/8887#issuecomment-1777754121_) posted by @rlh1994_

> Would it be useful to add a debug level warning when it fails to find a type? In particular if someone is using the `column.Data_type` from a `get_column_schema_from_query` call it might not be obvious why something is failing without directly outputting all the column data types?

_[Originally](https://github.com/dbt-labs/dbt-core/pull/8887/files#r1372209194) posted by @dbeatty10_

> @MichelleArk how would you feel about adding this ☝️ debug output when the `type_code` isn't found?

_[Originally](https://github.com/dbt-labs/dbt-core/pull/8887/files#r1372225638) posted by @MichelleArk_

> It may be a little redundant given `unknown type_code` will appear in the error message either way. So I'd either omit it or write a more informative message like"type_code not found in adapter-provided lookup, defaulting to 'unknown type_code'"
> 
> And we may actually need to use fire_event to fire a debuglevel event as opposed not logger.debug

The log level could be a warning rather than debug if we choose.

### Describe alternatives you've considered

A couple alternatives to `fire_event`:
1. `logger.debug`
1. No logging at all

### Who will this benefit?

The beneficiaries are similar to dbt-labs/dbt-core#8900. The primary intended audience is folks directly using `get_column_schema_from_query` (like described in https://github.com/dbt-labs/dbt-core/issues/8877) that might not realize a translation wasn't possible.

### Are you interested in contributing this feature?

PR for consideration: https://github.com/dbt-labs/dbt-core/pull/8957

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by dbeatty10**
resolves #38

### Problem

Postgres allows user defined data types. It also has some non-standard data types like `money`, `uuid`, etc. If dbt doesn't recognize the data type, dbt doesn't completely fail, but it also doesn't explain what's going on.

### Solution

Uses https://github.com/dbt-labs/dbt-adapters/pull/135 to emit a debug-level log message any time a data type is not recognized within dbt-postgres. That way, users can at least see in the logs what is going on.

Log messages will look like the following:

> The `type_code` 790 was not recognized

### 👉 After approval but before merge

- [ ] Run `git revert bc0b179` to undo pointing to dev branch in bc0b179

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-postgres/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests
- [x] This PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: pyproject.toml
Comment (mikealfare) [inline pyproject.toml:1]: Reminder to revert these changes once the PR in `dbt-adapters` is merged.

File: hatch_build.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 BASE_DEPS = [
     # psycopg2 dependency installed in custom hatch_build.py
-    "dbt-adapters>=0.1.0a1,<2.0",
+    "dbt-adapters @ git+https://github.com/dbt-labs/dbt-adapters.git@dbeatty/unknown-data-type-code-8912",
```
Comment (mikealfare) [inline hatch_build.py:10]: We need to revert this before merging. However, this causes an issue. If we don't point to GH, the integration tests will fail. If we do point to GH, then we can't release because PyPI doesn't not allow pointing to mutable versions (i.e. GH branches/tags). We need to resolve this in general (beyond the scope of this PR). I think for now, I would update this to point to `main` before merging and we'll discuss how to handle this in general prior to releasing.

File: dbt/adapters/postgres/connections.py [hunk 2]
```diff
@@ -203,4 +205,7 @@ def data_type_code_to_name(cls, type_code: Union[int, str]) -> str:
         if type_code in psycopg2.extensions.string_types:
             return psycopg2.extensions.string_types[type_code].name
         else:
+            warn_or_error(
```
Comment (mikealfare) [inline dbt/adapters/postgres/connections.py:208]: I'm surprised this had to be multiple lines; it must be just over the limit.

**Review by mikealfare — APPROVED** [review]
Looks good, just remember to revert the dependencies to point to `main`.

File: pyproject.toml
Comment (dbeatty10) [reply pyproject.toml:1]: 💯 

Yep! Have this as a checklist item in the PR, and will do that revert after dbt-adapters is merged and prior to merging to dbt-postgres

image

File: hatch_build.py (hunk 1 above)
Comment (mikealfare) [reply hatch_build.py:10]: We've since reverted the `hatch_build` approach, this should be back in `pyproject.toml` now.

**Comment by mikealfare** [comment]
I think this is going to be easier to merge post the `pre-commit` merge, so I'd like to get that in first and then rebase this PR. I added it to my sprint backlog for this sprint so I don't lose track of it.
